### DISCUSSION TITLE
Any way to set knitr::kable() as global option for all tables in a qmd file?

### DISCUSSION THREAD

**Discussion by jeweljohnsonj** [quarto-dev/quarto-cli#933] [Q&A] [answered]
Is there any way to set knitr::kable() as the default global option way of showing tables in a rendered file?

**Comment by jjallaire** [comment, marked as answer]
Here is one way to do it:

````markdown
```{r}
#| include: false
knit_print.data.frame = function(x, ...) {
  res = paste(c("", "", knitr::kable(x)), collapse = "\n")
  knitr::asis_output(res)
}

registerS3method(
  "knit_print", "data.frame", knit_print.data.frame,
  envir = asNamespace("knitr")
)
```

```{r}
mtcars
```
````

That's the low-level way, the `printr` package also implements this by default:

````markdown

```{r}
#| include: false
library(printr)
```

```{r}
mtcars
```
````

**Reply by jeweljohnsonj** [reply]
The `printr` package solves my problem. Thank you for the answers!

**Reply by mcanouil** [reply]
Are the other `Rmarkdown` methods to do this working in Quarto?

- `df_print` option
- `render` option

Source: https://bookdown.org/yihui/rmarkdown-cookbook/opts-render.html

**Reply by jeweljohnsonj** [reply]
Now I realize that knitr::kable() does not output compact and paged tables, like how rmarkdown::paged_table() output does. Would be nice if I could have paged_table() as the default way of printing.

@mcanouil I tried searching if `df_print` and `render` option is there for quarto in this [link](https://quarto.org/docs/output-formats/html-basics.html), it doesn't seem its there.

**Reply by mcanouil** [reply]
Indeed, I did not find them as well, just checking since related to the question ^^

**Reply by cderv** [reply]
`render` is a **knitr** option so it should be available even within Quarto but this is an advanced option, and not all functions are working for `render`. The **printr** package is heping with that by defining method wrapping function (like `printr:::knit_print.data.frame` wrapping `knitr::kable()` so that it can be used as a render function. 

There are several way to do what you need above. But first let's answer the other question : `df_print` is different as this is really a R Markdown feature supported using a chunk option. There is no equivalent option in Quarto. 

However, you can recreate the feature for your document. One can just use what is shown in the R Markdown Cookbook - the render function just needs to be resulting in what **knitr** expects. 

For paged table this would be this function: (which is used internaly by **rmarkdown** for the `df_print` feature.
````r
paged_print <- function(x, options) {
  knitr::asis_output(
    rmarkdown:::paged_table_html(x, options = attr(
      x,
      "options"
    )),
    meta = list(dependencies = rmarkdown::html_dependency_pagedtable())
  )
}
````

Then you could register it in a document so that it is applied by default for any data.frame printing.
````r
registerS3method(
  "knit_print", "data.frame", paged_print,
  envir = asNamespace("knitr")
)
````

or use it on chunk basis where you need to print a data.frame in a single value chunk.
````r
```{r, render = paged_print}
iris
```
````

A `df_print` chunk option can also be simulated this way using a [option hook ](https://bookdown.org/yihui/rmarkdown-cookbook/option-hooks.html)

````r
knitr::opts_hooks$set(df_print = function(options) {
  if (options$df_print == "paged") {
    options$render = paged_print
  }
  options
})
````

this will allow something like 
````markdown
```{r, df_print = "paged"}
iris
```

```{r}
iris
```
````
* First table will be shown as paged table
* Second table will be show as usual data.frame

This would work for R output only. So I don't know if we want to make this internal to Quarto. If we do add paged table feature, then it will be differently. 

I'll think more about this, but I believe we could include such feature in the **printr** package or another place. 

Hope the above helps already achieve what you need.

**Reply by mcanouil** [reply]
Thanks for this detailed answer!

My take on the subject: I'd rather have/keep something general in Quarto and use S3 R methods or `knitr` hooks, if/when needed.

**Reply by cderv** [reply]
> I'd rather have/keep something general in Quarto and use S3 R methods or knitr hooks, if/when needed.

I agree with that. Maybe not the right place but I throw the idea in **printr** (https://github.com/yihui/printr/issues/41) we'll think about it.

**Reply by jeweljohnsonj** [reply]
Thank you so much for the answer!

**Comment by cderv** [comment]
For using with R, there is now a `df-print` option you can set in YAML: https://quarto.org/docs/computations/r.html#data-frames

So you can now set `kable` as default printing option.

**Reply by jeweljohnsonj** [reply]
Thanks for adding the feature.
